rl4co/models/nn/attention.py

```diff
@@ -220,6 +211,53 @@ def _inner_mha(self, query, key, value, attn_mask):
     def _make_heads(self, v):
         return rearrange(v, "... g (h s) -> ... h g s", h=self.num_heads)
 
+    def _project_out(self, out):
+        return self.project_out(out)
+
+
+class PointerAttnMoE(PointerAttention):
+    """Calculate logits given query, key and value and logit key.
+    This follows the pointer mechanism of Vinyals et al. (2015) (https://arxiv.org/abs/1506.03134),
+        and the MoE gating mechanism of Zhou et al. (2024) <https://arxiv.org/abs/2405.01029>.
+
+    Note:
+        With Flash Attention, masking is not supported
+
+    Performs the following:
+        1. Apply cross attention to get the heads
+        2. Project heads to get glimpse
+        3. Compute attention score between glimpse and logit key
+
+    Args:
+        embed_dim: total dimension of the model
+        num_heads: number of heads
+        mask_inner: whether to mask inner attention
+        linear_bias: whether to use bias in linear projection
+        check_nan: whether to check for NaNs in logits
+        sdpa_fn: scaled dot product attention function (SDPA) implementation
+        moe_kwargs: Keyword arguments for MoE
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        num_heads: int,
+        mask_inner: bool = True,
+        out_bias: bool = False,
+        check_nan: bool = True,
+        sdpa_fn: Optional[Callable] = None,
+        moe_kwargs: Optional[dict] = None,
+    ):
+        super(PointerAttnMoE, self).__init__(embed_dim, num_heads, mask_inner, out_bias, check_nan, sdpa_fn)
+        self.moe_kwargs = moe_kwargs
+
+        if self.moe_kwargs is not None:
+            self.project_out = None
+            self.project_out_moe = MoE(embed_dim, embed_dim, num_neurons=[], out_bias=out_bias, **moe_kwargs)
+            if self.moe_kwargs["light_version"]:
+                self.dense_or_moe = nn.Linear(embed_dim, 2, bias=False)
+                self.project_out = nn.Linear(embed_dim, embed_dim, bias=out_bias)
+
     def _project_out(self, out):
         """Implementation of Hierarchical Gating based on Zhou et al. (2024) <https://arxiv.org/abs/2405.01029>."""
         if self.moe_kwargs is not None:
```

Comment:
(read comment below) if `PointerAttnMoe` by default has `MoE`, then we could also remove this "if, else" statement